File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    max_dungeons = 0 # 탐험할 수 있는 최대 던전수
    answer = -1
    
    # 모든 던전 탐험 순서에 대해 시도
    for perm in permutations(dungeons):
        current_tired = k # 현재 남은 피로도를 초기 피로도 설정
        dungeons_explored = 0 # 탐험한 던전 수를 세기 위한 변수 초기화
        
        # 현재 순서의 던전들에 대해 탐험 시도
        for min_required, tired_cost in perm:
            # 현재 피로도가 던전의 최소 필요 피로도 이상일때만 탐헌 가능
            if current_tired >= min_required:
                current_tired -= tired_cost # 던전 탐험 -> 피로도 소모
                dungeons_explored += 1 # 던전 탐험 수 증가
            else:
                break
        # 최대 던전 수 업데이트
        max_dungeons = max(max_dungeons, dungeons_explored)
    return max_dungeons

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    private int maxExplored;
    
    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        
        boolean[] visited = new boolean[dungeons.length]; // 방문 배열
        recur(k, dungeons, visited, 0);
        
        return maxExplored;
    }
    
    private void recur(int k, int[][] dungeons, boolean[] visited, int cnt ){
        
        // 탐험
        for(int i=0; i<dungeons.length; i++){
            // 방문 x && 현재 피로도 >= 최소 피로도
            if(!visited[i] && k>= dungeons[i][0]){
                // 방문 체크
                visited[i]=true;
                // 현재 피로도, 방문체크, 탐험횟수+1 갱신한 재귀 함수 실행
                recur(k-dungeons[i][1], dungeons, visited,cnt+1 );
                // 실행 후 방문체크 되돌려 다른 순서 탐험 가능하도록.
                visited[i]=false;
            }
            // 최댓값 구하기
            maxExplored = Math.max(cnt, maxExplored);
        }
    }
}